test/extended: Add MultiNetworkPolicy test case

zeeke opened the pull request:
The test case uses Cluster Network Operator field `useMultiNetworkPolicy` field and a basic `deny-all` policy to ensure the feature is correctly setup.
Fixture involves a macvlan net-attach-def and network policy. Connectivity tests are implemented using agnhost http server and cURL.

Signed-off-by: Andrea Panattoni <apanatto@redhat.com>

zeeke commented:
/retest-required

danwinship commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+		err = oc.AsAdmin().Run("create").Args("-f", nad_yaml).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("launching pod with an annotation to use the net-attach-def")
```
We're probably way overdue for some NAD-related helper functions...

```
err = exutil.CreateNetworkAttachmentDefinition("macvlan-nad.yml")

exutil.SetNADAnnotation(pod, "macvlan1-nad", map[string]string{
        "ips": fmt.Sprintf(`["%s"]`, ip),
})
```

etc?

danwinship commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+
+	o.Eventually(func() error {
+		return oc.AsAdmin().Run("get").Args("multi-networkpolicies.k8s.cni.cncf.io").Execute()
+	}, "30s", "2s").Should(o.Succeed())
```
This doesn't reliably indicate that MNP _enforcement_ is in effect... that depends on deploying a DaemonSet as well, right?

The timeouts help, but it would be better if there was a good way of ensuring that the DaemonSet is fully deployed, so the test doesn't end up flaking under heavy load. (But also, we don't want to make too many assumptions about exactly what CNO is doing when you enable the feature, especially since we might change to a different MNP implementation in the future... I'm not sure what the best approach here is...)

zeeke replied:
I agree with your considerations, it's not easy to decide what's the best approach. From the user's perspective, turning on the feature means that MNP resource becomes available. 

If the Daemonset (or any other component) is not working or it is not started, then some other assertion will fail.

As a further example, in the future we can also decide to keep the Daemonset up and running but not creating iptables if the flag is set to false.

That said, I would keep it as simple as possible.

danwinship commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+		err = oc.AsAdmin().Run("create").Args("-f", multinetpolicy_yaml).Execute()
+		o.Expect(err).To(o.Succeed())
+
+		g.By("checking podB can NOT connnect to podA")
```
I think you need to test more than this. As written, this test would pass even if the actual behavior was "activating MNP completely breaks all secondary networks" :grimacing:

zeeke replied:
Good point! To achieve it, I need to:
- add a third pod (podC)
- change the policy from a deny-all style to something like:
```
...
  podSelector:
    matchLabels:
      pod: a
  policyTypes:
    - Ingress
  ingress:
    - from:
        - podSelector:
            matchLabels:
              pod: c
```
- check pod-C can connect to pod-A

But it will make the test a little more complicated. Do you have any suggestions for this?

danwinship commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+	}, "30s", "2s").Should(o.Succeed())
+}
+
+func disablMultiNetworkPolicy(oc *exutil.CLI) {
```
typo "disabl"

danwinship commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+		output, err := oc.AsAdmin().Run("exec").Args(testPodB.Name, "--", "curl", "--connect-timeout", "1", "2.2.2.1:8889").Output()
+		o.Expect(err).To(o.Succeed(), "cmd output: %s", output)
+
+		g.By("enabling MultiNetworkPolicies on cluster")
+		enableMultiNetworkPolicy(oc)
+		defer disablMultiNetworkPolicy(oc)
```
You're assuming that the feature is disabled by default in all clusters where the e2e suite runs, but that doesn't seem safe. I'd check to see if it's already enabled, and only do the "enable and defer disable" if it's not.

zeeke replied:
Sure, I added logic to store initial `useMultinetworkPolicy` state and restore it afterward

danwinship commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+		err = oc.AsAdmin().Run("create").Args("-f", nad_yaml).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("launching pod with an annotation to use the net-attach-def")
```
um, wait, there's already a `createNetworkAttachmentDefinition` method in `test/extending/networking/utils.go`

zeeke replied:
Yep, saw it .But it uses dynamic client instead of Yaml files. Is it ok to turn every NAD creation (I found 4 in the code base) in a call to `createNetworkAttachmentDefinition()`? and then removing yaml files?

zeeke commented on test/extended/networking/multinetpolicy.go:
```diff
@@ -0,0 +1,102 @@
+			]`
+
+		frameworkpod.CreateExecPodOrFail(f.ClientSet, ns, podName, func(pod *v1.Pod) {
+			pod.Spec.Containers[0].Args = []string{"net", "--serve", "2.2.2.1:8889"}
```
Cool, I didn't know that: https://www.rfc-editor.org/rfc/rfc5737 .

zeeke commented:
/retest-required

zeeke commented:
@danwinship can you please take another look at this?

zeeke commented:
/retest-required

cgoncalves commented:
/retest-required

zeeke commented:
/retest-required

cgoncalves commented:
/retest-required

zeeke commented:
@cgoncalves 
@danwinship 

All the required tests are passing. `openshift-ci` comments refer to old tests that are no longer required.

- https://github.com/openshift/release/pull/33550

e.g. `e2e-gcp-builds` -> `e2e-gcp-ovn-builds`

cgoncalves commented:
/lgtm

zeeke commented:
@s1061123 @dougbtv @danwinship 
can you help me get this in?

s1061123 reviewed (changes requested):
Almost stuff seems to be good. Thanks!

zeeke commented on test/extended/testdata/multinetpolicy/deny-ingress-pod-a.yml:
```diff
@@ -0,0 +1,11 @@
+  annotations:
+    k8s.v1.cni.cncf.io/policy-for: macvlan1-nad
+spec:
+  podSelector:
```
sure!

zeeke commented:
@s1061123 thanks for reviewing. Since this issue staled for a while, I opened an IPv6 / IPv4 version in:
- https://github.com/openshift/origin/pull/27795

I addressed your comment there. Please, have a look

zeeke commented:
Closing in favor of:
- https://github.com/openshift/origin/pull/27795
